core/src/main/java/org/apache/calcite/sql/SqlPeriodTypeNameSpec.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.sql;
+
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.sql.validate.SqlValidator;
+import org.apache.calcite.util.Litmus;
+
+import static org.apache.calcite.util.Static.RESOURCE;
+
+/**
+ * A sql type name specification of the PERIOD type.
+ */
+public class SqlPeriodTypeNameSpec extends SqlTypeNameSpec {
+
+  public final TimeScale timeScale;
+  public final Integer precision;
+  public final boolean isWithTimezone;
+
+  public SqlPeriodTypeNameSpec(TimeScale timeScale,
+      SqlNumericLiteral precision,
+      boolean isWithTimezone, SqlParserPos pos) {
+    super(new SqlIdentifier("Period", pos), pos);
+
+    // date period cannot contain precision or with time zone token
+    if (timeScale == TimeScale.DATE
+        && (precision != null || isWithTimezone)) {
+      throw SqlUtil.newContextException(pos,
+          RESOURCE.illegalNonQueryExpression());
+    }
+
+    if (precision != null) {
+      if (!precision.isInteger()) {
+        throw SqlUtil.newContextException(pos,
+            RESOURCE.illegalNonQueryExpression());
+      }
+      int precisionValue = Integer.parseInt(precision.toValue());
+      if (precisionValue < 0 || precisionValue > 6) {
+        throw SqlUtil.newContextException(pos,
+            RESOURCE.numberLiteralOutOfRange("Precision"));
+      }
+      this.precision = precisionValue;
+    } else {
+      this.precision = null;
```

Comment:
nit: If I'm not mistaken, this.precision will default to null if not set.

Reply:
Good catch. It should default to 6 if not set for TIME and TIMESTAMP. Fixed in https://github.com/googleinterns/calcite/pull/177/commits/110d83e563426a9ea05113263c6efc02c44447dc

Reply:
I think my original comment here still holds: do we even need to explicitly set this.precision to null here?

Reply:
I think we need to explicitly set this.precision to null. This final variable will not be initialize in the else case if we don't do this step. If we instead set it to be null at the beginning, then this.precision can't be a final variable.

Reply:
IIUC, this documentation suggests that we do not need to explicitly set precision to null. Instead, it will automatically inherit that value if it isn't set to something else.

https://en.wikipedia.org/wiki/Uninitialized_variable#:~:text=Java%20does%20not%20have%20uninitialized,null%20for%20all%20reference%20types).

Reply:
If this is the case, I'd suggest just removing this explicit setting to null.

Reply:
See the following wiki reference on [java final variable](https://en.wikipedia.org/wiki/Final_(Java)#Final_variables).  Which it says "A blank final instance variable of a class must be definitely assigned in every constructor of the class in which it is declared;". Intellij also recognized there would be an error if I declare a final variable as a blank final but never assign it a value.

Reply:
Ah, I didn't realize that - good to know. Ok then I agree it would be best to leave it as final and explicitly set to null in that case.